### DISCUSSION TITLE
Help on serving OpenPose to Triton backend

### DISCUSSION THREAD

**Discussion by JJLimmm** [triton-inference-server/server#5766] [General]
Similar to this issue #3648 , i wanted to ask if anybody managed to use Openpose with triton backend for inference.
I am planning to use the Body_25 model, but it is in caffe and that's not one of the supported backends for Triton... 

Or has anybody managed to convert the caffe model from Openpose to ONNX/Pytorch/Tensorflow?

**Comment by tanmayv25** [comment]
Does this Tensorflow model resolve your issue: https://modelzoo.co/model/lightweight-openpose ?

**Reply by JJLimmm** [reply]
I saw that repository before, but was wondering how to include the loading of images to feed into the model and to post process the outputs for using the different frameworks (ONNX/Pytorch/Tensorflow). 

Currently, OpenPose is written in C++ and it has a python Wrapper for loading the model, preprocessing and post processing. However, i'm not sure how to include these steps to be part of the triton inference server since the model is automatically loaded by triton instead of loading it as part of the openpose wrapper if i were to use the model for inference without triton. (i.e: on client device, it sends an image. On triton inference server, it will load the model, then receive the image, use the python wrapper to preprocess, somehow send that preprocessed data to the model for inference with the python wrapper, and then post process the outputs again with the wrapper.)

**Reply by tanmayv25** [reply]
@JJLimmm If it is difficult to convert the OpenPose model to one of the Triton supported frameworks, you can write the serving in a python script and use Triton's Python backend to run the caffe model in Triton.
https://github.com/triton-inference-server/python_backend

**Reply by JJLimmm** [reply]
@tanmayv25 I saw this Python backend serving on triton as well. A little confused at their documentation. Not sure if i can just containerize the whole project in a docker image and run in with triton or i need to write a python script to start the triton inference.

**Reply by tanmayv25** [reply]
You can containerize the whole project. See the simple example here: https://github.com/triton-inference-server/python_backend/tree/main/examples/pytorch

Python backend allows users to deploy a python script as a model in Triton Inference Server. 

The overall workflow would look like:

`Triton Client -> Triton Server Endpoint -> Triton Core -> Triton Python Backend -> model.py`

In your case the solution should look like:


`Triton Client -> Triton Server Endpoint -> Triton Core -> Triton Python Backend -> model.py -> caffe model`

**Reply by JJLimmm** [reply]
I see, so i will have to write out the _**TritonPythonModel()**_ class.
Am i right if i do this below?

1. write initialize() method in TritonPythonModel() class to load the openpose caffe model via their Openpose Python Wrapper.
2. write the execute() method for handling the inference request, but via the OpenPose Wrapper.

For the second part, am i even still able to perform the inference request without using pb_utils library API (python backend utils)? 
Or do you know of how i can still use the Openpose python wrapper within the _**pb_utils.InferenceResponse()**_ API call?

**Reply by tanmayv25** [reply]
This looks accurate. You'd have to obtain the output as numpy array. Then create pb_utils.Tensor object from this numpy array and feed these pb_utils.Tensor objects as output tensors to pb_utils.InferenceResponse().

I am assuming Openpose python wrapper does provide results in form of numpy array, or in a format that can be converted to numpy.
